caas/kubernetes/provider/utils/labels.go

```diff
@@ -4,52 +4,131 @@
 package utils
 
 import (
-	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"context"
+
+	"github.com/juju/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	core "k8s.io/client-go/kubernetes/typed/core/v1"
 
 	"github.com/juju/juju/caas/kubernetes/provider/constants"
 )
 
-func LabelSetToSelector(labels k8slabels.Set) k8slabels.Selector {
-	return k8slabels.SelectorFromValidatedSet(labels)
-}
+var (
+	// LabelsJuju is a common set
+	LabelsJuju = map[string]string{
+		constants.LabelKubernetesAppManaged: "juju",
+	}
+)
 
-// AppendLabels adds the labels defined in src to dest returning the result.
-// Overlapping keys in sources maps are overwritten by the very last defined
-// value for a duplicate key.
-func AppendLabels(dest map[string]string, sources ...map[string]string) map[string]string {
-	if dest == nil {
-		dest = map[string]string{}
+// HasLabels returns true if the src contains the labels in has
+func HasLabels(src, has labels.Set) bool {
+	for k, v := range has {
+		if src[k] != v {
+			return false
+		}
 	}
-	if sources == nil {
-		return dest
+	return true
+}
+
+// IsLegacyModelLabels checks to see if the provided model is running on an older
+// labeling scheme or a newer one.
+func IsLegacyModelLabels(model string, namespaceI core.NamespaceInterface) (bool, error) {
+	ns, err := namespaceI.Get(context.TODO(), model, meta.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return false, nil
 	}
-	for _, s := range sources {
-		for k, v := range s {
-			dest[k] = v
-		}
+	if err != nil {
+		return true, errors.Annotatef(err, "unable to determine legacy status for namespace %s", model)
 	}
-	return dest
+
+	return !HasLabels(ns.Labels, LabelsForModel(model, false)), nil
+}
+
+// LabelSetToSelector converts a set of Kubernetes labels
+func LabelSetToSelector(l labels.Set) labels.Selector {
+	return labels.SelectorFromValidatedSet(l)
 }
 
 // LabelsForApp returns the labels that should be on a k8s object for a given
 // application name
-func LabelsForApp(name string) map[string]string {
-	return map[string]string{
-		constants.LabelApplication: name,
+func LabelsForApp(name string, legacy bool) labels.Set {
```

Comment:
If we have a SelectorLabelsForApp() as well, we can avoid the boilerpate used in many places to add the LabelsJuju